### ISSUE TITLE
Basic Linear Model with mask fails Autoregression test

### ISSUE THREAD

**Issue by amogkam** [pytorch/pytorch#17457]
## 🐛 Bug

A simple 2 layer linear model with the correct masks from https://arxiv.org/pdf/1502.03509.pdf fails the autoregressive property tests. The input gradient is 0 even when there is a connection from the input to the output.

## To Reproduce

```
import numpy as np

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch import tensor
from collections import OrderedDict

class MaskedLinear(nn.Linear):
	""" same as Linear except has a configurable mask on the weights """

	def __init__(self, in_features, out_features, bias=True):
		super().__init__(in_features, out_features, bias)
		self.register_buffer('mask', torch.ones(out_features, in_features))

	def forward(self, input):
		return F.linear(input, self.mask * self.weight, self.bias)

if __name__ == '__main__':
	rng = np.random.RandomState(14)

	nin = 3
	hidden_size = 4
	nout = 5
	input_bins = [2, 1, 2]

	net = [MaskedLinear(nin, hidden_size), nn.ReLU(), MaskedLinear(hidden_size, nout)]
	model = nn.Sequential(*net)

	d = OrderedDict([('0.weight', tensor([[-0.1124, -0.4300,  0.2062],
        [ 0.1824, -0.4534,  0.3814],
        [-0.4321, -0.4623,  0.3500],
        [ 0.2166, -0.2858, -0.4061]])), ('0.bias', tensor([-0.4513, -0.4768, -0.4137,  0.3184])), ('0.mask', tensor([[1., 1., 0.],
        [1., 1., 0.],
        [1., 0., 0.],
        [1., 1., 0.]])), ('2.weight', tensor([[-0.1391,  0.3570,  0.2720,  0.3102],
        [ 0.2908,  0.2073, -0.2786, -0.0651],
        [-0.1322,  0.3701, -0.2332,  0.4623],
        [ 0.4570,  0.1008, -0.3818,  0.1526],
        [-0.2794, -0.2873, -0.1177,  0.2994]])), ('2.bias', tensor([-0.3249, -0.4900, -0.0787, -0.3565, -0.4509])), ('2.mask', tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 1., 0.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]]))])

	model.load_state_dict(d)
	model.eval()

	k = 2
	inp = torch.tensor(
			rng.rand(1, nin).astype(np.float32), requires_grad=True)

	loss = model(inp)
	l = loss[0, k]
	l.backward()

	print('grad = ', inp.grad)

	assert inp.grad[0, 0] != 0 and inp.grad[0, 1:] == 0 
```

## Expected behavior

On each iteration, the output and index k should only depend on the the previous inputs. Those gradients should be non-zero while all the other gradients should be 0. However, for k=2, the gradients for all 3 inputs are 0 even though the first input should have a non-zero gradient. This fails the autoregressive property even though the masks are correct.

## Environment
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.6) 5.4.0 20160609
CMake version: version 3.5.2

Python version: 3.5
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.13.3
[pip3] numpydoc==0.7.0
[pip3] torch==1.0.1.post2
[conda] mkl                       2018.0.0             hb491cac_4
[conda] mkl-service               1.1.2            py35h0fc7090_4
[conda] torch                     1.0.1.post2               <pip>

**Comment by concretevitamin** [comment]
In other words, for a certain 2-layer MLP, some output unit should propagate gradients back to inputs (say) `0, 1, 2` but the gradients there are zeros.

**Comment by amogkam** [comment]
Here is a diagram of what the neural net looks like with the masks applied. 
![untitled document](https://user-images.githubusercontent.com/8068268/53317104-3a340300-387f-11e9-9d83-4fb156896062.jpg)

When k=2 in the code, this is referring to output `o_1_0` in the diagram. There is a connection from this output to input `in_0`, but not the other 2. Thus, the gradient wrt `in_0` should be non-zero, but the gradients wrt `in_1` and `in_2` should be 0. However, when printing the gradient, it is 0 for all three.

`grad =  tensor([[0., 0., 0.]])`

**Comment by soumith** [comment]
@amogkam why did you close the issue? any conclusion would be helpful for readers that visit this link at a later time.

**Comment by concretevitamin** [comment]
Relu pushes some values to zeros so they get a gradient of 0 back.  The autoregressive check of checking gradient != 0 was too strict.
